psi4/driver/qcdb/vib.py

```diff
@@ -0,0 +1,1062 @@
+        herm = np.allclose(a, a.conj().T, atol=atol)
+        ivrt = a.shape[0] - np.linalg.matrix_rank(a, tol=stol)
+        return """  {:32} Symmetric? {}   Hermitian? {}   Lin Dep Dim? {:2}""".format(lbl + ':', symm, herm, ivrt)
+
+    def vec_in_space(vec, space, tol=1.0e-4):
+        merged = np.vstack((space, vec))
+        u, s, v = np.linalg.svd(merged)
+        return (s[-1] < tol)
+
+    vibinfo = {}
+    text = []
+
+    nat = len(mass)
+    text.append("""\n\n  ==> Harmonic Vibrational Analysis <==\n""")
+
+    if nat == 1:
+        nrt_expected = 3
+    elif np.linalg.matrix_rank(geom) == 1:
+        nrt_expected = 5
+    else:
+        nrt_expected = 6
+
+    nmwhess = hess.copy()
+    text.append(mat_symm_info(nmwhess, lbl='non-mass-weighted Hessian') + ' (0)')
+
+    # get SALC object, possibly w/o trans & rot
+    mints = core.MintsHelper(basisset)
+    cdsalcs = mints.cdsalcs(0xFF, project_trans, project_rot)
+
+    Uh = collections.OrderedDict()
+    for h, lbl in enumerate(irrep_labels):
+        tmp = np.asarray(cdsalcs.matrix_irrep(h))
+        if tmp.size > 0:
+            Uh[lbl] = tmp
+
+    # form projector of translations and rotations
+    space = ('T' if project_trans else '') + ('R' if project_rot else '')
+    TRspace = _get_TR_space(mass, geom, space=space, tol=LINEAR_A_TOL)
+    nrt = TRspace.shape[0]
+    text.append(
+        '  projection of translations ({}) and rotations ({}) removed {} degrees of freedom ({})'.
+        format(project_trans, project_rot, nrt, nrt_expected))
+
+    P = np.identity(3 * nat)
+    for irt in TRspace:
+        P -= np.outer(irt, irt)
+    text.append(mat_symm_info(P, lbl='total projector') + ' ({})'.format(nrt))
+
+    # mass-weight & solve
+    sqrtmmm = np.repeat(np.sqrt(mass), 3)
+    sqrtmmminv = np.divide(1.0, sqrtmmm)
+    mwhess = np.einsum('i,ij,j->ij', sqrtmmminv, nmwhess, sqrtmmminv)
+    text.append(mat_symm_info(mwhess, lbl='mass-weighted Hessian') + ' (0)')
+
+    pre_force_constant_au = np.linalg.eigvalsh(mwhess)
+
+    idx = np.argsort(pre_force_constant_au)
+    pre_force_constant_au = pre_force_constant_au[idx]
+    uconv_cm_1 = np.sqrt(psi_na * psi_hartree2J * 1.0e19) / (2 * np.pi * psi_c * psi_bohr2angstroms)
+    pre_frequency_cm_1 = np.lib.scimath.sqrt(pre_force_constant_au) * uconv_cm_1
+
+    pre_lowfreq = np.where(np.real(pre_frequency_cm_1) < 100.0)[0]
+    pre_lowfreq = np.append(pre_lowfreq, np.arange(nrt_expected))  # catch at least nrt modes
+    for lf in set(pre_lowfreq):
+        vlf = pre_frequency_cm_1[lf]
+        if vlf.imag > vlf.real:
+            text.append('  pre-proj  low-frequency mode: {:9.4f}i [cm^-1]'.format(vlf.real, vlf.imag))
+        else:
+            text.append('  pre-proj  low-frequency mode: {:9.4f}  [cm^-1]'.format(vlf.real, ''))
+    text.append('  pre-proj  all modes:' + str(_format_omega(pre_frequency_cm_1, 4)))
+
+    # project & solve
+    mwhess_proj = np.einsum('ij,jk,kl->il', P.T, mwhess, P)
```

Comment:
This is very elegant machinery, and I like the details being provided to the user the way they are

Reply:
Hmm, careful here without `np.einsum` optimize machinery this could scale like N^4. Might be better write `np.dot(P.T, mwhess).dot℗`. This counteracts Andy, but that could actually become a bottleneck!

Reply:
Good point. Back to `.dot`